presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ApplyConnectorOptimization.java

```diff
@@ -292,4 +300,26 @@ private static <T> boolean containsAll(Set<T> container, Collection<T> test)
         }
         return true;
     }
+
+    private static class ConnectorToInternalJoinRewriter
+            extends SimplePlanRewriter<Void>
+    {
+        @Override
+        public PlanNode visitConnectorJoinNode(ConnectorJoinNode node, RewriteContext<Void> context)
```

Comment:
Maybe I'm missing something obvious - but why do we need to have a new type `ConnectorJoinNode` ? Why can't we move the `JoinNode` type to the SPI ?

Reply:
When I tried I got stuck on a lot of classes that needed to be moved as well. But I don't remember which ones at the moment, I could give it another go. However there is also some value in having a dedicated class acting as an adapter to shield the SPI from internal changes.

Reply:
I think it was `FunctionAndTypeManager` which isn't available to the SPI. I guess we can add a helper class that is available to the callers of `getSortExpressionContext`. 
There were more classes but most of them needed to be moved to the SPI anyway for the adapter node. 

It's just a one liner.

Reply:
Okay I think this is a fair compromise